google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkValidator.java

```diff
@@ -0,0 +1,133 @@
+import com.google.cloud.tools.appengine.cloudsdk.CloudSdk;
+import com.google.cloud.tools.appengine.cloudsdk.CloudSdkNotFoundException;
+import com.google.cloud.tools.appengine.cloudsdk.CloudSdkOutOfDateException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.intellij.openapi.components.ServiceManager;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/** Helper service to check validity of {@link CloudSdkService}. */
+public class CloudSdkValidator {
+
+  public static CloudSdkValidator getInstance() {
+    return ServiceManager.getService(CloudSdkValidator.class);
+  }
+
+  protected Set<com.google.cloud.tools.intellij.appengine.sdk.CloudSdkValidationResult>
+      validateCloudSdk(Path path) {
+    Set<com.google.cloud.tools.intellij.appengine.sdk.CloudSdkValidationResult> validationResults =
+        new HashSet<>();
+
+    if (path == null) {
+      validationResults.add(
+          com.google.cloud.tools.intellij.appengine.sdk.CloudSdkValidationResult
+              .CLOUD_SDK_NOT_FOUND);
+      // If the Cloud SDK is not found, don't bother checking anything else
+      return validationResults;
+    }
+
+    CloudSdk sdk = buildCloudSdkWithPath(path);
+    try {
+      sdk.validateCloudSdk();
+    } catch (CloudSdkNotFoundException exception) {
+      validationResults.add(
+          com.google.cloud.tools.intellij.appengine.sdk.CloudSdkValidationResult
+              .CLOUD_SDK_NOT_FOUND);
+      // If the Cloud SDK is not found, don't bother checking anything else
+      return validationResults;
+    } catch (CloudSdkOutOfDateException exception) {
+      validationResults.add(
+          com.google.cloud.tools.intellij.appengine.sdk.CloudSdkValidationResult
+              .CLOUD_SDK_VERSION_NOT_SUPPORTED);
+    }
+
+    try {
+      sdk.validateAppEngineJavaComponents();
+    } catch (AppEngineJavaComponentsNotInstalledException ex) {
+      validationResults.add(
+          com.google.cloud.tools.intellij.appengine.sdk.CloudSdkValidationResult
+              .NO_APP_ENGINE_COMPONENT);
+    }
+
+    return validationResults;
+  }
+  /** Checks if the default SDK stored path contains a valid Cloud SDK. */
+  public Set<com.google.cloud.tools.intellij.appengine.sdk.CloudSdkValidationResult>
+      validateCloudSdk() {
+    return validateCloudSdk(CloudSdkService.getInstance().getSdkHomePath());
+  }
+
+  /**
+   * Checks if a given path is malformed or if it contains a valid Cloud SDK.
+   *
+   * <p>Windows' implementation of Paths doesn't handle well converting strings with certain special
+   * characters to paths. This method should be called before {@code Paths.get(path)}.
+   */
+  public Set<com.google.cloud.tools.intellij.appengine.sdk.CloudSdkValidationResult>
```

Comment:
why the fully qualified name?

Reply:
Oops, some refactoring glitch, cleaned up.